### PR TITLE
Symbol table test

### PR DESCRIPTION
**Pull request by S3JER**
Added test to the RecSymbolTable class and rewrote the RecSymbolTable OpenScope method.

### CODE REVIEW

File: GOAT-Compiler/RecSymbolTable.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+using System.Threading.Tasks;
+
+namespace GOAT_Compiler
+{
+    class RecSymbolTable : ISymbolTable
+    {
+
+        private Table tables;
+        private Table mainTable;
+        private bool buildFlag = false;
```
Comment (Mati-AAU) [inline GOAT-Compiler/RecSymbolTable.cs:14]: This seems poorly named. As i read it,  if "buildflag" is false, it is building, which seems wrong

File: GOAT-Compiler/RecSymbolTable.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace GOAT_Compiler
+{
+    class RecSymbolTable : ISymbolTable
+    {
+
+        private Table tables;
```
Comment (Mati-AAU) [inline GOAT-Compiler/RecSymbolTable.cs:12]: what is "tables"?

File: GOAT-Compiler/RecSymbolTable.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        private Table mainTable;
```
Comment (Mati-AAU) [inline GOAT-Compiler/RecSymbolTable.cs:13]: maybe call it globaltable cause its the global scope

File: GOAT-Compiler/RecSymbolTable.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            mainTable = new Table(tables);
+            tables.ChildrenTables.Add(mainTable);
+            currentScope = tables;
+        }
+
+        public void OpenScope()
+        {
+            if (currentScope.ParentTable == null)
+            {
+                currentScope = currentScope.ChildrenTables[0];
```
Comment (Mati-AAU) [inline GOAT-Compiler/RecSymbolTable.cs:29]: this is hard to understand, a quick comment?

File: GOAT-Compiler/RecSymbolTable.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            currentScope.VisitCounter = 0;
+            currentScope = currentScope.ParentTable;
+        }
+
+        public Symbol GetSymbol(string Name)
+        {
+            return currentScope.GetSymbol(Name);
+        }
+
+        public void AddSymbol(string Name, Types type)
```
Comment (Mati-AAU) [inline GOAT-Compiler/RecSymbolTable.cs:64]: Maybe this should take a symbol as a parameter instead?

File: GOAT-Compiler/RecSymbolTable.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        {
+            currentScope.SetSymbol(Name, type);
+        }
+
+    }
+
+    class Table
```
Comment (Mati-AAU) [inline GOAT-Compiler/RecSymbolTable.cs:71]: access modifier for table? internal probably?

File: GOAT-Compiler/RecSymbolTable.cs [hunk 7]
```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace GOAT_Compiler
+{
+    class RecSymbolTable : ISymbolTable
```
Comment (Mati-AAU) [inline GOAT-Compiler/RecSymbolTable.cs:9]: access modifier?

File: GOAT-Compiler/RecSymbolTable.cs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            }
+            else if(ParentTable != null)
+            {
+                return ParentTable.GetSymbol(Name);
+            }
+
+            return null;
+        }
+
+        public void SetSymbol(string Name, Types type)
```
Comment (Mati-AAU) [inline GOAT-Compiler/RecSymbolTable.cs:101]: maybe this should take a symbol as parameter instead?

File: SymbolTableTest/IntegrationTest.cs [hunk 9]
```diff
@@ -0,0 +1,110 @@
+using System;
+using Xunit;
+using GOAT_Compiler;
+
+namespace SymbolTableTest
+{
+    public class IntegrationTest
```
Comment (Mati-AAU) [inline SymbolTableTest/IntegrationTest.cs:7]: integration test isnt a fitting name anymore, also the name should relate to symbol table

File: SymbolTableTest/IntegrationTest.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+            symbolTable.OpenScope();
+            symbolTable.AddSymbol(value, type);
+            Assert.Equal(type, symbolTable.GetSymbol(value).type);
+            symbolTable.CloseScope();
+        }
+
+        [Theory]
+        [InlineData("a", 2, Types.Integer)]
+        [InlineData("b", 3, Types.Boolean)]
+        public void Checks_Acces_After_Build(string symbolName, int depth, Types type)
```
Comment (Mati-AAU) [inline SymbolTableTest/IntegrationTest.cs:25]: typo, access not acces

File: SymbolTableTest/IntegrationTest.cs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            for (int i = 1; i <= depth; i++)
+            {
+                symbolTable.CloseScope();
+            }
+            symbolTable.CloseScope();
+        }
+
+        [Theory]
+        [InlineData("a", "a", Types.Boolean)]
+        public void Check_Doublicate_Decl_In_SymbolTable(string firestSymbol, string secondSymbol, Types type)
```
Comment (Mati-AAU) [inline SymbolTableTest/IntegrationTest.cs:54]: duplicate not doublicate

File: SymbolTableTest/IntegrationTest.cs [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+            symbolTable.OpenScope();
+            symbolTable.AddSymbol("a", Types.Integer);
+            symbolTable.OpenScope();
+            Assert.NotNull(symbolTable.GetSymbol("a"));
+            symbolTable.CloseScope();
+            symbolTable.CloseScope();
+        }
+
+        [Fact]
+        public void Check_If_GetSymbol_Returns_Null()
```
Comment (Mati-AAU) [inline SymbolTableTest/IntegrationTest.cs:85]: returns null in what situation? not said in name

File: SymbolTableTest/IntegrationTest.cs [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+            symbolTable.OpenScope();
+            symbolTable.OpenScope();
+            symbolTable.AddSymbol("a", Types.Integer);
+            symbolTable.CloseScope();
+            Assert.Null(symbolTable.GetSymbol("a"));
+            symbolTable.CloseScope();
+        }
+
+        [Fact]
+        public void Test()
```
Comment (Mati-AAU) [inline SymbolTableTest/IntegrationTest.cs:97]: this one is just called test :p

**Review by Mati-AAU — CHANGES_REQUESTED** [review]
overall p cool but cleanup required

File: SymbolTableTest/IntegrationTest.cs (hunk 9 above)
Comment (S3JER) [reply SymbolTableTest/IntegrationTest.cs:7]: Renamed it to SymbolTableTest

File: SymbolTableTest/IntegrationTest.cs (hunk 10 above)
Comment (S3JER) [reply SymbolTableTest/IntegrationTest.cs:25]: fixed

File: SymbolTableTest/IntegrationTest.cs (hunk 11 above)
Comment (S3JER) [reply SymbolTableTest/IntegrationTest.cs:54]: Fixed

File: SymbolTableTest/IntegrationTest.cs (hunk 12 above)
Comment (S3JER) [reply SymbolTableTest/IntegrationTest.cs:85]: Renamed it to Check_Outer_Scope_Access_If_Not_Found_Return_Null

File: SymbolTableTest/IntegrationTest.cs (hunk 13 above)
Comment (S3JER) [reply SymbolTableTest/IntegrationTest.cs:97]: It is named Chekcs_The_Right_Symbol_Gets_Returned_When_In_A_Scope now

File: GOAT-Compiler/RecSymbolTable.cs (hunk 2 above)
Comment (Casper-NS) [reply GOAT-Compiler/RecSymbolTable.cs:12]: The tables contain all other tables

File: GOAT-Compiler/RecSymbolTable.cs (hunk 3 above)
Comment (Casper-NS) [reply GOAT-Compiler/RecSymbolTable.cs:13]: Yes Sir!
Renamed to globalScope

File: GOAT-Compiler/SymbolTable.cs [hunk 14]
```diff
@@ -0,0 +1,144 @@
+﻿using System;
```
[the remark this answers is not in the document]
Comment (Casper-NS) [reply GOAT-Compiler/SymbolTable.cs]: Will fix in next branch

File: GOAT-Compiler/ISymbolTable.cs [hunk 15]
```diff
@@ -0,0 +1,13 @@
+﻿namespace GOAT_Compiler
+{
+    internal interface ISymbolTable
```
[the remark this answers is not in the document]
Comment (Casper-NS) [reply GOAT-Compiler/ISymbolTable.cs]: Will fix in next branch

**Pull request merged**